Method for exposing the raw data

paalge opened the pull request:
This is a method for exposing the raw data as a numpy array.
Not sure if this is exactly the way you want to do it, but it looks like it is working for me

Fixes #47

SamWhited commented on rawkit/raw.py:
```diff
@@ -7,6 +7,7 @@
 import random
 import string
 import tempfile
+import numpy as np
```
We definitely don't want a hard dependency on numpy. Can we use a regular array for now, and we'll figure out a good general API to support numpy (hopefully w/o copying data to Python land first) later?

paalge replied:
I normally use numpy, so I have very little experience with the regular array. As long as you can write to a normal uint16 array from an iterator numpy isn't needed.

SamWhited replied:
Yah, I just meant a normal Python array. It won't be very efficient, but I've also opened an issue (#106) for discussing the best way to handle numpy support, which is something we want to do.

paalge replied:
II understand. Can you read from an iterator into such an array?

SamWhited replied:
Yup, that should be fine. You won't be able to use the numpy methods to do it though; just regular iteration will work fine.

SamWhited commented on rawkit/raw.py:
```diff
@@ -175,6 +176,52 @@ def save_thumb(self, filename=None):
+
+        # Raise error if 4 color image isn't there, which will happen for some
+        # cameras
+        raise_if_error(rawdata.color4_image.contents == ctypes.c_voidp)
```
`raise_if_error` was really made to be an internal (to the `libraw` package) helper method that deals with error codes. It shouldn't be used here. I'll open an issue to possibly update it to make this clearer, sorry about that.

paalge replied:
I just used it as a default. There needs to be a check here as `color4_image` contains `Null`for images that don't have 4 colours (as for instance jpg)

SamWhited replied:
Yah, I just meant that a regular check (not this method) will work fine. This method actually has logic that's not valid for the comparison, it just happens to work because of the way Python type massages booleans.

```python
if rawdata.color4_image.contents is None:
    return None
```

will probably work (I'm not actually sure that we want this to be an error; since there just isn't any raw data, returning None makes sense to me). Then again, I'm not srue what would make a better API, maybe we need to make a `NoDataError` or something. @campaul?

paalge commented:
I've removed the numpy dependency and changed the error handling

SamWhited commented on rawkit/raw.py:
```diff
@@ -175,6 +175,57 @@ def save_thumb(self, filename=None):
+
+        try:
+            rawdata.color4_image.contents
+        except:
```
Ah, this makes sense, thanks! Let's catch the actual exception here though; bare excepts can have unintended side effects :)

SamWhited commented on rawkit/raw.py:
```diff
@@ -175,6 +175,57 @@ def save_thumb(self, filename=None):
+        )
+
+        # make 2D list
+        data = [[0 for i in range(iheight)] for j in range(iwidth)]
```
Might be cleaner, no idea if there's a "standard" way to do this in Python:

```python
data = [None] * iheight * iwidth
```

It might be a pre-mature optimization either way though.

paalge replied:
It needs to be 2D, or you'll have to pass height and width on, and let the calling function do the reshape. 
From what I understood, list compression is the normal way of initialising this.

SamWhited replied:
Sounds good then. I think the above will make a 2d list too, but I could be mistaken.

paalge replied:
A quick quick test gives it as a 1D. 
As for the preallocate, I like it due to the large lists being created, though I don't know what is fastest.

SamWhited replied:
If we're going to loop and append, 0's though we might as well just loop and append the data. I'm not sure this is actually an optimization.

campaul replied:
This would probably be more efficient. I'll do some timings later and see for sure.
```
data = [[] for j in range(iwidth)]

for ii in range(iheight):
    for jj in range(iwidth):
        data[jj].append(data_pointer[ii * iwidth + jj])
```

SamWhited commented on rawkit/raw.py:
```diff
@@ -175,6 +175,58 @@ def save_thumb(self, filename=None):
+
+        for ii in range(iheight):
+            for jj in range(iwidth):
+                data[jj][ii] = data_pointer[ii * iwidth + jj]
```
This still appears to be subscripting a 2D array; won't this be broken?

paalge replied:
In my test run this works as intended. 
Though I'm not sure what you think would be broken?
Line 221 makes a 2D list, which one can index with [][] in line 225

SamWhited replied:
Oh right, yah, I was misunderstanding a large chunk of our discussion, sorry. I think it's best to make this a 1D list to match the data that we get from LibRaw / elsewhere in rawkit. Not sure though; will consider the API implications of doing this.

campaul commented on rawkit/raw.py:
```diff
@@ -175,6 +175,58 @@ def save_thumb(self, filename=None):
+        # make 2D list
+        data = [[0 for i in range(iheight)] for j in range(iwidth)]
+
+        for ii in range(iheight):
```
Can we change `ii` and `jj` to `y` and `x` respectively so we're working with Cartesian coordinates?

SamWhited replied:
These aren't actually coordinates on the image though since there are multiple indexes per pixel.

paalge replied:
They will be coordinates, though you have to interpolate the image first, as in that loop they change between different colour channels.

SamWhited commented on rawkit/raw.py:
```diff
@@ -175,6 +175,58 @@ def save_thumb(self, filename=None):
         self.libraw.libraw_dcraw_thumb_writer(
             self.data, filename.encode('ascii'))
 
+    def get_4_col_raw(self):
```
Another thing I'm thinking (though this doesn't have to be done right now by any means, I just wanted to comment while I was thinking about it so that it was documented somewhere) is that the public facing API method for this should just work for all types of raws (regardless of how many channels there are). @campaul and I will probably sit down sometime this weekend and figure out a more unified API (including return values and types) for all methods that return image data like this. Will advise.

paalge replied:
Hi.

It is easy to change this to the other case (with only 3 colours), as the data then is in `color3_image` and not in `color4_image`. Though before I modify this I would like to know what the return types should be. In my mind we should return something that is 2D, as this removes the need to pass along height and width.

campaul commented:
@paalge if you go ahead and add a `get_3_col_raw` function, update the docs to say the functions are experimental, and then squash your commits I'm happy merging this.

We'll be making some changes to the interface as time goes on, but I want to get the core functionality merged in so I can start playing with it.

paalge commented:
@campaul I've now added the requested method. I also fixed a lot of error in the original code. The returned image is the image excluding the calibration pixels (it could be possible to switch between this in an input flag). I've only tested the code one three types of images (ARW, NEF and three colour CR2). The three colour image is strange, as it includes a zero pixel.
Can you have a look at the code? I'll squash the git when you've taken a look

campaul commented:
In general the code looks good, but the two functions are almost identical. Can you extract the shared code into a helper function so we aren't duplicating everything?

paalge commented:
I'm sorry for the late update. I've now moved the common code into a new function, though I'm unsure of how much better it is with respect to readability, as there where some subtle differences in the code.

campaul commented:
Oops, I didn't see the email from github for your last comment. I had already taken your commits and made the changes I wanted, as well as updated the tests, in #108. Could you take a look at that branch and let me know if you're still ok with how the code looks? If you're happy with things I'll merge that branch in and we should be good to go.

campaul commented:
This was included as part of #108.
